pytorch_lightning/trainer/callback_hook.py

```diff
@@ -175,3 +176,22 @@ def on_keyboard_interrupt(self):
         """Called when the training is interrupted by KeyboardInterrupt."""
         for callback in self.callbacks:
             callback.on_keyboard_interrupt(self, self.get_model())
+
+    def on_save_checkpoint(self):
+        """Called when saving a model checkpoint."""
+        callback_states = {}
+        for callback in self.callbacks:
+            callback_class = type(callback)
+            state = callback.on_save_checkpoint(self, self.get_model())
+            if state:
+                callback_states[callback_class] = state
+        return callback_states
+
+    def on_load_checkpoint(self, checkpoint):
+        """Called when loading a model checkpoint."""
+        callback_states = checkpoint.get('callbacks')
+        for callback in self.callbacks:
+            state = callback_states.get(type(callback))
+            if state:
+                state = deepcopy(state)
+                callback.on_load_checkpoint(state)
```

Comment:
I'm wondering, should we call on_load_checkpoint even if there is no state, i.e., pass empty dict? Maybe the user has a callback in which the state is meaningful only sometimes, and there is some code that needs to run always. I have no concrete example but just thinking of the edge cases here.

Reply:
@jeremyjordan Actually, i take it back. I think your previous version was good. Consider the following use case:

We first train a model but our callbacks=[model_checkpoint] (no early stopping)
It saved a checkpoint. Now we slightly change the code to 
callbacks==[model_checkpoint, early_stopping] (we added ES). 
We resume training with our new code. This will get a KeyError in the on_load_checkpoint because it tries to access values that are not in the checkpoint.

Reply:
@jeremyjordan Actually, i take it back. I think your previous version was good. Consider the following use case:

We first train a model but our callbacks=[model_checkpoint] (no early stopping)
It saved a checkpoint. Now we slightly change the code to
callbacks==[model_checkpoint, early_stopping] (we added ES).
We resume training with our new code. This will get a KeyError in the on_load_checkpoint because it tries to access values that are not in the checkpoint.

Reply:
ah, great point! will revert :)